### PR TITLE
[Fix] S2-Pro argmax decoding degeneration causing vocoder OOM (#235)

### PR DESCRIPTION
**Pull request by yxs**
**Root cause**: The CUDA-graphed inference path uses `torch.argmax` for semantic token selection, while fish-speech upstream uses `torch.multinomial` sampling. With argmax, if `im_end` never becomes the highest-logit token for a given input, the model never stops, it runs to `max_new_tokens` every time. The oversized codebook sequence then OOMs the vocoder.

  **Fix**: Replace `torch.argmax` with `torch.multinomial` sampling (temperature=0.7, matching fish-speech upstream default). `torch.multinomial` is CUDA-graph compatible.

  Tested on H100 80GB, SeedTTS EN 1088 (`--no-ref-audio`, concurrency=1):

  |  | Before | After |
  |---|---|---|
  | Pass | 1079/1088 | 1088/1088 |
  | Fail | 9 (same 9 every run) | 0 |
  | Latency mean | 1.293s | 1.26s |
  | WER (seedtts-50, whisper-large-v3) | 1.45% | 0.91% |

### CODE REVIEW

File: sglang_omni/models/fishaudio_s2_pro/sglang_model.py [hunk 1]
```diff
@@ -272,6 +273,7 @@ def setup_vq_decode(
         bias[semantic_begin_id : semantic_end_id + 1] = 0.0
         bias[im_end_id] = 0.0
         self._semantic_bias = bias
```
[the remark this answers is not in the document]
Comment (yxs) [reply sglang_omni/models/fishaudio_s2_pro/sglang_model.py:274]: overprotect

**Comment by zhaochenyang20** [comment]
> The CUDA-graphed inference path uses torch.argmax for semantic token selection, while fish-speech upstream uses torch.multinomial sampling. 

Does this means that enable or not CUDA Graph would go into two different decoding path, and CUDA Graph somehow hurts the accuracy?

https://github.com/sgl-project/sglang-omni/issues/235

As you mentioned here, there are 9 cases failing consistently. We 

```bash
9/1088 requests fail consistently. Failing IDs:
  - common_voice_en_18436934-common_voice_en_18436935 — "Biotechnology, or Biotech, is a very popular degree choice"
  - common_voice_en_18466098-common_voice_en_18466097 — "Not a thing."
  - common_voice_en_20048255-common_voice_en_20048259 — "Queen who were both fighting for control of the city."
  - common_voice_en_20410042-common_voice_en_20410055 — "New World China Land Limited is wholly owned by the Group."
  - common_voice_en_37232389-common_voice_en_37190783 — "Will you be happy when I am in the earth?"
  - common_voice_en_485101-common_voice_en_485100 — "I'll always let you down!"
  - common_voice_en_523215-common_voice_en_523219 — "She needed to be pretty observant to have noticed such a small crack."
  - common_voice_en_597578-common_voice_en_597583 — "Wilma's gone cat sitting again for the weekend if you want to come over."
  - common_voice_en_63703-common_voice_en_63702 — "Now go upstairs and see what little Christina wants."
  ```

File: tests/test_model/test_s2pro_tts_ci.py [hunk 2]
```diff
@@ -378,6 +380,49 @@ def test_voice_cloning_streaming_consistency(
         )
 
 
+# ---------------------------------------------------------------------------
```
Comment (zhaochenyang20) [inline tests/test_model/test_s2pro_tts_ci.py]: remove these lines `-----------`

```python

Note (Xuesong): xxx
```

File: tests/test_model/test_s2pro_tts_ci.py [hunk 3]
```diff
@@ -378,6 +380,49 @@ def test_voice_cloning_streaming_consistency(
...
+                json={"input": text},
+                timeout=60,
+            )
+        assert resp.status_code == 200, (
+            f"Degenerate text #{i + 1} failed: {resp.text[:200]}"
+        )
+        assert len(resp.content) > 1000, (
+            f"Degenerate text #{i + 1} produced too little audio: "
+            f"{len(resp.content)} bytes"
+        )
```
Comment (zhaochenyang20) [inline tests/test_model/test_s2pro_tts_ci.py]: change the dataset of `download_dataset(DATASETS["seedtts-50"], str(root), quiet=True)`

**Review by zhaochenyang20 — APPROVED** [review]

**Comment by zhaochenyang20** [comment]
1. Change the dataset of Seed TTS 50. Add nine hard samples.
2. Evaluate with the current main; we should have a rather worse WER on the new dataset.
3. Apply yours, then we should have a better WER compared with 2.

Codes are clean. Great. Fix lint, please.

**Comment by zhaochenyang20** [comment]
I have some strong feedback so far. We can PR it in two steps. This PR could do 1 and change the dataset.

Thanks for the fix — the root cause analysis is solid and the benchmark numbers are convincing. Two pieces of feedback:

## 1. Simplify the sampling fix

The core issue is that the CUDA Graph path incorrectly replaced `torch.multinomial` with `torch.argmax`. The fix should just be reverting to `torch.multinomial` to match fish-speech upstream — no need to introduce a new `semantic_temperature` parameter, the extra `torch.softmax` path, or the plumbing into `setup_vq_decode`.

Right now this PR adds unnecessary API surface (`semantic_temperature` kwarg, `self._semantic_temperature` attribute), which then triggers follow-up concerns (Copilot's `temperature <= 0` validation comment, the full-vocab softmax perf note). All of that goes away if we just do:

```python
# Before (broken)
semantic_token = torch.argmax(biased_logits, dim=-1)

# After (match upstream)
probs = torch.softmax(biased_logits / 0.7, dim=-1)
semantic_token = torch.multinomial(probs, num_samples=1).squeeze(-1)
```

If we want temperature to be configurable later, that's a separate PR.

## 2. Vocoder memory guard (follow-up needed)

This PR fixes the common case (sampling prevents degenerate decode), but the underlying vulnerability remains: **there is no validation that `max_new_tokens` is within what the vocoder can actually handle given available GPU memory.**

Normal requests finish in 50–300 tokens. `max_new_tokens=2048` produces a `[10, 2048]` codebook sequence that needs ~1.5 GiB in the DAC conv1d layers — enough to OOM an H100 after model + KV cache are loaded. If any future bug causes AR to hit the max again, vocoder will OOM again.

What we should do (as a follow-up issue/PR):
- At server launch, pre-compute the maximum sequence length the vocoder can support given the actual available GPU memory after model loading.
- If the user's configured `max_new_tokens` exceeds this limit, log a warning and clamp it down to the safe maximum.
- This way sampling fixes the 99% case, and the vocoder guard covers the remaining edge cases.

## Minor

- Remove the `# -----------` separator and `Note (Xuesong)` comment block in the test file.
- Fix lint (already flagged by CI).
- Per your earlier comment: add the 9 hard samples to seedtts-50 dataset and show before/after WER — agreed, this would make the CI improvement concrete.

Overall: approve the direction, requesting the simplification above before merge.

**Comment by yxs** [comment]
@zhaochenyang20 
Will do.
FYI, the token distribution from 1088 benchmark: P95: 150, P99: 307, Max: 613

File: tests/test_model/test_s2pro_tts_ci.py (hunk 3 above)
Comment (yxs) [reply tests/test_model/test_s2pro_tts_ci.py]: `xuesongye/seed-tts-eval-59`

File: tests/test_model/test_s2pro_tts_ci.py [hunk 4]
```diff
@@ -378,6 +380,44 @@ def test_voice_cloning_streaming_consistency(
...
+            )
+        assert (
+            resp.status_code == 200
+        ), f"Degenerate text #{i + 1} failed: {resp.text[:200]}"
+        assert len(resp.content) > 1000, (
+            f"Degenerate text #{i + 1} produced too little audio: "
+            f"{len(resp.content)} bytes"
+        )
+
+
```
Comment (zhaochenyang20) [inline tests/test_model/test_s2pro_tts_ci.py]: remove this.

File: sglang_omni/models/fishaudio_s2_pro/sglang_model.py [hunk 5]
```diff
@@ -349,9 +349,10 @@ def _decode_codebooks(self, logits: Tensor, hidden_states: Tensor) -> None:
         """Constrained semantic sampling + batched codebook generation."""
         bs = logits.shape[0]
 
-        # Constrained decode: mask non-semantic tokens
+        # Constrained decode: mask non-semantic tokens, then sample
         biased_logits = logits + self._semantic_bias
-        semantic_token = torch.argmax(biased_logits, dim=-1)  # [bs]
+        probs = torch.softmax(biased_logits / 0.7, dim=-1)
```
Comment (zhaochenyang20) [inline sglang_omni/models/fishaudio_s2_pro/sglang_model.py:356]: add a comment:

```python

Note (Xuesong): Following Non-CUDA Graph xxx
# reference: https://github.com/sgl-project/sglang-omni/pull/267
```

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Comment by yxs** [comment]
WER threshold from 2.5 -> 1.5

Rounds 1–20: 0.91%, 1.27%, 0.91%, 1.27%, 1.27%, 1.45%, 0.91%, 1.27%, 1.09%, 1.09%, 0.91%, 0.91%, 1.09%, 0.91%, 0.91%, 1.27%, 0.91%, 1.27%, 1.09%, 1.09%
Min: 0.91% | Max: 1.45% | Mean: 1.09%

Streaming: 1.27%

File: sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py [hunk 6]
```diff
@@ -334,8 +334,48 @@ def _get_stream_codec() -> Any:
...
+      # Requests whose max_new_tokens exceed this limit are clamped and raise
+      # warnings.
+
+      # Caveat: PyTorch's caching allocator gradually consumes free memory
+      # over time, so actual headroom at runtime is lower than at startup.
+      # This guard is conservative but not airtight.
+
+      # reference: https://github.com/sgl-project/sglang-omni/pull/267
+
+    _VOCODER_BYTES_PER_TOKEN = int(5.3 * 1024 * 1024)
```
Comment (zhaochenyang20) [inline sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py]: Put the const at the begining of file.

File: sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py [hunk 7]
```diff
@@ -334,8 +334,48 @@ def _get_stream_codec() -> Any:
...
+    logger.info(
+        "Vocoder memory guard: GPU free %.1f GB, max_vocoder_tokens=%d",
+        free_mem / 1e9,
+        max_vocoder_tokens,
+    )
+
     def _request_builder(payload: StagePayload):
         state = load_state(payload)
+        if state.max_new_tokens > max_vocoder_tokens:
+            logger.warning(
```
Comment (zhaochenyang20) [inline sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py:373]: use f-string for logging. Do not use `%s`

**Comment by yxs** [comment]
H100
fishaudio/s2-pro served via `sglang_omni.cli.cli serve`

- Server startup allocation: 74,641 MiB
- Free memory after startup: 6,449 MiB (~6.4 GB)
- Guard computes: max_vocoder_tokens = 6449 / 5.3 ≈ 1216

| Experiment | Requests | Drift | Observation |
|---|---|---|---|
| 100 diverse texts | 100 | +5,592 MiB | Stabilized at ~request 80, last 20 deltas all 0 |
| Same text 30x (graphs ON) | 30 | +18 MiB | Near-zero growth |
| Same text 30x (graphs OFF) | 30 | +2,790 MiB | Significant growth even with identical input |
| Similar-length different-content (25 req) | 25 | +2,230 MiB | Same input length still causes growth |
| 8 concurrent requests × 2 batches | 16 | +10 MiB | Concurrency does not worsen growth |

1. The caching allocator reserves ~5.5 GB over ~100 diverse requests then stabilizes. This memory is reusable — the server successfully processed all 100 requests (including 1.8 MB audio outputs) with only 857 MiB nvidia-smi free remaining.
2. CUDA graphs dramatically reduce cache growth (same text: +18 MiB vs +2,790 MiB without graphs).
3. When the allocator needs a large contiguous block, it evicts cached blocks back to CUDA. Observed: nvidia-smi free *increased* by 832 MiB between consecutive requests during a short→long text test.
4. The startup-based guard is a reasonable upper bound because the vocoder allocates through the same PyTorch allocator and can reuse/evict cached blocks on demand.

File: sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py [hunk 8]
```diff
@@ -334,8 +336,44 @@ def _get_stream_codec() -> Any:
...
+        f"Vocoder memory guard: GPU free {free_mem / 1e9:.1f} GB, max_vocoder_tokens={max_vocoder_tokens}"
+    )
+
     def _request_builder(payload: StagePayload):
         state = load_state(payload)
+        if state.max_new_tokens > max_vocoder_tokens:
+            logger.warning(
+                f"Request {payload.request_id}: max_new_tokens={state.max_new_tokens} exceeds vocoder limit {max_vocoder_tokens}, clamping."
+            )
+            state.max_new_tokens = max_vocoder_tokens
```
[the remark this answers is not in the document]
Comment (yxs) [reply sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py:376]: not reasonable

File: sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py
Comment (zhaochenyang20) [inline sglang_omni/models/fishaudio_s2_pro/pipeline/stages.py:332]: duplicated with `    gpu_id_int = int(device.split(":")[-1]) if ":" in device else 0`.

File: sglang_omni/models/fishaudio_s2_pro/sglang_model.py [hunk 9]
```diff
@@ -349,9 +349,12 @@ def _decode_codebooks(self, logits: Tensor, hidden_states: Tensor) -> None:
         """Constrained semantic sampling + batched codebook generation."""
         bs = logits.shape[0]
 
-        # Constrained decode: mask non-semantic tokens
+        # Constrained decode: mask non-semantic tokens, then sample
         biased_logits = logits + self._semantic_bias
-        semantic_token = torch.argmax(biased_logits, dim=-1)  # [bs]
+        # Note (Xuesong): Following Non-CUDA Graph path with temperature=0.7 (fish-speech upstream default).
```
Comment (zhaochenyang20) [inline sglang_omni/models/fishaudio_s2_pro/sglang_model.py:354]: I still found that the non-graph path (`runtime/s2pro_ar.py`'s `_sample_with_topk`) includes top-k=30, top-p, repetition penalty, and per-request temperature, while the CUDA graph path only does `softmax(logits / 0.7) + multinomial` — no top-k, no top-p, no repetition penalty, hardcoded temperature.

Is this intented?

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: sglang_omni/models/fishaudio_s2_pro/sglang_model.py (hunk 9 above)
Comment (yxs) [reply sglang_omni/models/fishaudio_s2_pro/sglang_model.py:354]: Yes, intentional as a fix for the argmax degeneration OOM. The hardcoded 0.7 follows the fish-speech upstream default.
Current WER passes and audio quality is fine.
Full sampling alignment can be a follow-up PR, though implementing top-k/top-p inside CUDA graphs adds complexity. I'll experiment with it and see how it goes.

Comment (zhaochenyang20) [reply sglang_omni/models/fishaudio_s2_pro/sglang_model.py:354]: Good. Please open a issue to track this full alignment.

Comment (yxs) [reply sglang_omni/models/fishaudio_s2_pro/sglang_model.py:354]: https://github.com/sgl-project/sglang-omni/issues/272

**Review by zhaochenyang20 — APPROVED** [review]
Good to merge after CI. Let me merge it please.
